### PR TITLE
[SPARK-30592][SQL] Interval support for csv and json funtions

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

In this PR, I'd propose to fully support interval for the CSV and JSON functions. 

On one hand, CSV and JSON records consist of string values, in the cast logic, we can cast string from/to interval now, so we can make those functions support intervals easily.

Before this change we can only use this as a workaround.
```sql
SELECT cast(from_csv('1, 1 day', 'a INT, b string').b as interval)
struct<CAST(from_csv(1, 1 day).b AS INTERVAL):interval>
1 days
```

On the other hand,  we ban reading or writing intervals from CSV and JSON files. To directly read and write  with external json/csv storage, you still need explicit cast, e.g.
```scala
spark.read.schema("a string").json("a.json").selectExpr("cast(a as interval)").show
+------+
|     a|
+------+
|1 days|
+------+
```

### Why are the changes needed?

for interval's future-proofing purpose

### Does this PR introduce any user-facing change?

yes, the `to_json`/`from_json` function can deal with intervals now. e.g.
for `from_json` there is no such use case because we do not support `a interval` 
for `to_json`, we can use interval values now 

#### before

 ```sql

 SELECT to_json(map('a', interval 25 month 100 day 130 minute));
Error in query: cannot resolve 'to_json(map('a', INTERVAL '2 years 1 months 100 days 2 hours 10 minutes'))' due to data type mismatch: Unable to convert column a of type interval to JSON.; line 1 pos 7;
'Project [unresolvedalias(to_json(map(a, 2 years 1 months 100 days 2 hours 10 minutes), Some(Asia/Shanghai)), None)]
+- OneRowRelation
```
#### after
```sql
SELECT to_json(map('a', interval 25 month 100 day 130 minute))
{"a":"2 years 1 months 100 days 2 hours 10 minutes"}

```
### How was this patch tested?

add ut

### CODE REVIEW

**Review by MaxGekk — COMMENTED** [review]
Could you add a test for `spark.sql.legacy.to_json.verifyMapKeyType.enabled` is `false` to be sure that the config works.

**Comment by yaooqinn** [comment]
Thanks, @MaxGekk on second thought, we should support these functions to deal with intervals as we only need to prohibit interval from external storage. So I have removed that conf

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala [hunk 1]
```diff
@@ -680,25 +678,20 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
       output)
   }
 
-  test("to_json: verify MapType's value type instead of key type") {
-    // Keys in map are treated as strings when converting to JSON. The type doesn't matter at all.
-    val mapType1 = MapType(CalendarIntervalType, IntegerType)
-    val schema1 = StructType(StructField("a", mapType1) :: Nil)
-    val struct1 = Literal.create(null, schema1)
+  test("from_json - interval support") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala]: nit: can we follow csv test and test from/to in one test case?

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 2]
```diff
@@ -222,33 +222,24 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
       Row("""{"_1":"26/08/2015 18:00"}""") :: Nil)
   }
 
-  test("to_json - key types of map don't matter") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: I think we can remove it now. There is no unsupported types anymore. JSON supports all the types.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala]: sgtm

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 3]
```diff
@@ -252,6 +252,12 @@ class JacksonParser(
           Decimal(bigDecimal, dt.precision, dt.scale)
       }
 
+    case CalendarIntervalType => (parser: JsonParser) =>
+      parseJsonToken[CalendarInterval](parser, dataType) {
+        case VALUE_STRING =>
+          IntervalUtils.safeStringToInterval(UTF8String.fromString(parser.getText))
+    }
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: nit: shift `}` to the right, it belongs to `parseJsonToken` not to `case`

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: thanks

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 4]
```diff
@@ -220,3 +220,9 @@ select a - b from values (interval '-2147483648 months', interval '2147483647 mo
 select b + interval '1 month' from values (interval '-2147483648 months', interval '2147483647 months') t(a, b);
 select a * 1.1 from values (interval '-2147483648 months', interval '2147483647 months') t(a, b);
 select a / 0.5 from values (interval '-2147483648 months', interval '2147483647 months') t(a, b);
+
+-- interval support for csv and json functions
+SELECT from_csv('1, 1 day', 'a INT, b interval');
+SELECT to_csv(named_struct('a', interval 32 month, 'b', interval 70 minute));
+SELECT from_json('{"a":"1 days"}', 'a interval');
+SELECT to_json(map('a', interval 25 month 100 day 130 minute));
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql]: @yaooqinn Could you add round-trip test cases?

- from_csv(to_csv(....))
- to_csv(from_csv(...))
- from_json(to_json(....))
- to_json(from_json(....))
-

**Comment by gatorsmile** [comment]
@yaooqinn `Does this PR introduce any user-facing change?` does not include the changes made in CSV. We need to update the PR description when you updated the code.

**Comment by yaooqinn** [comment]
> @yaooqinn `Does this PR introduce any user-facing change?` does not include the changes made in CSV. We need to update the PR description when you updated the code.

The changes in CSV are not user-facing

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql (hunk 4 above)
Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql]: ok, I'll make a followup.

**Pull request closed without merging**